### ISSUE TITLE
Mr. Upbeat is very inconsistent with starting up.

### ISSUE THREAD

**Issue by blank3times** [RHeavenStudio/HeavenStudio#522]
### Describe the bug

When Mr. Upbeat switches in from another game and you prepare at or after the switch block, or "When the switch block isn't at beat 0, and the prepare block is after", Mr. Upbeat tends not to start at all.

This can inconsistently be remedied by a Ding! block, but then it doesn't blip anymore the next time you restart the remix when the box is ticked. It feels like the ding is somehow affecting the play afterwards, like how in the past (and maybe present), Karate Man's background wouldn't reset to default upon starting a remix.

I've been experiencing heavy inconsistency, where sometimes it does start without blipping, or sometimes doesn't start, or when I move the block from before (to test that it still works and it does) to after, it starts blipping before the first time and then stops again after.

### How to reproduce the bug?

Various positions of Mr. Upbeat's switch game, prepare, and ding blocks that make sense, that do not include:

- Switch game on the first beat (that's normal)
- Prepare before switch game (that's normal)
- Ding before any preparation (that's not reasonable)
- Multiple of anything in a row (that's not needed)

You can also put a game before the Mr. Upbeat if you want.

Play the remix a few times, maybe move the blocks around. (I hope you can reproduce it?)

You can move the blocks to working positions (prepare before switch) and play it if you move it back and play it again after for adverse results (if you are actually able to reproduce).

### What Operating System(s) were you using?

Windows 10/11

### Log file

```shell
Mono path[0] = 'C:/Users/[]/HEAVEN STUDIO/Heaven Studio_Data/Managed'
Mono config path = 'C:/Users/[]/HEAVEN STUDIO/MonoBleedingEdge/etc'
Initialize engine version: 2021.3.21f1 (1b156197d683)
[Subsystems] Discovering subsystems at path C:/Users/[]/HEAVEN STUDIO/Heaven Studio_Data/UnitySubsystems
GfxDevice: creating device client; threaded=1; jobified=1
Direct3D:
    Version:  Direct3D 11.0 [level 11.1]
    Renderer: Intel(R) HD Graphics 520 (ID=0x1916)
    Vendor:   Intel
    VRAM:     10150 MB
    Driver:   30.0.100.9865
Begin MonoManager ReloadAssembly
- Completed reload, in  0.484 seconds
<RI> Initializing input.

XInput1_3.dll not found. Trying XInput9_1_0.dll instead...
<RI> Input initialized.

<RI> Initialized touch support.

UnloadTime: 4.491300 ms
No JoyShocks found.
The character used for Underline is not available in font asset [rodin_lat_cy_ja_ko_spec SDF].
Unloading 6 Unused Serialized files (Serialized files now loaded: 0)
UnloadTime: 10.774500 ms
Running game loader RvlBadmintonLoader
Running game loader CtrBearLoader
Running game loader RvlBoardMeetingLoader
Running game loader NtrFlickLoader
Running game loader CtrCatchLoader
Running game loader RvlBookLoader
Running game loader AgbClapLoader
Running game loader NtrCoinLoader
Running game loader NtrCropLoader
Running game loader NtrDjLoader
Running game loader NtrDogNinjaLoader
Running game loader RvlDoubleDateLoader
Running game loader CtrDrummingLoader
Running game loader NtrIdolLoader
Running game loader AgbFireworkLoader
Running game loader CtrFirstContact
Running game loader RvlFlipperFlopLoader
Running game loader RvlForkLoader
Running game loader NtrGleeClubLoader
Running game loader RvlNewKarateLoader
Running game loader CtrTeppanLoader
Running game loader RvlRocketLoader
Running game loader NtrBackbeatLoader
Running game loader AgbMarcherLoader
Running game loader pcoMeatLoader
Running game loader AgbUpbeatLoader
Running game loader ntrMunchyMonkLoader
Running game loader NtrOctopusMachineLoader
Running game loader CtrPillowLoader
Running game loader AgbQuizShowLoader
Running game loader NtrPingpongLoader
Running game loader PcoSomenLoader
Running game loader AgbHairLoader
Running game loader RvlRingsideLoader
Running game loader NtrRockersLoader
Running game loader NtrSamuraiLoader
Running game loader RvlSeeSawLoader
Running game loader AgbGhostLoader
Running game loader AgbSpaceDanceLoader
Running game loader NtrSoccerLoader
Running game loader AgbBatterLoader
Running game loader RvlTambourineLoader
Running game loader AgbTapLoader
Running game loader RvlTapTroupeLoader
Running game loader NtrDazzlesLoader
Running game loader AgbTossBoysLoader
Running game loader AgbTramLoader
Running game loader MobTrickLoader
Running game loader NtrTunnelLoader
Running game loader AgbWaltzLoader
Running game loader RvlWorkingDoughLoader
Discord status updated
Create TimingDisplayComponent
Create SkillStarComponent
Create SectionComponent
Unloading 595 unused Assets to reduce memory usage. Loaded Objects now: 7223.
Total: 14.448500 ms (FindLiveObjects: 0.706200 ms CreateObjectMapping: 0.476300 ms MarkObjects: 12.528700 ms  DeleteObjects: 0.736300 ms)

Update Success!
Imported RIQ successfully!
Detected "v1" riq (RiqBeatmapData)
Running beatmap update handlers
Preprocessing beatmap...
Running entity update handlers
ASYNC loading assetbundles for game mrUpbeat
== Playthrough statistics of New Remix (played at 2023-08-08 00:06:03) ==
Average input offset for playthrough: 0ms
Accuracy for playthrough:  0.00
Cleared 0 sections out of 0
Skill Star not collected
ASYNC loading assetbundles for game mrUpbeat
HeavenStudio
chart has no music: Chart song file does not exist at path C:/Users/[]/AppData/Local/Temp/RHeavenStudio/Heaven Studio/RIQCache/song.bin
Discord status updated
Update Success!
Playing at 0
Preloading game mrUpbeat
Event not found! May be spelled wrong or it is not implemented.
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
  at System.Collections.Generic.List`1[T].get_Item (System.Int32 index) [0x00009] in <88e4733ac7bc4ae1b496735e6b83bbd3>:0 
  at HeavenStudio.Util.MultiSound.Play (HeavenStudio.Util.MultiSound+Sound[] snds, System.Boolean game, System.Boolean forcePlay) [0x00020] in <6d2f5506bdb644ce89fc7390a1ea8702>:0 
  at HeavenStudio.Games.MrUpbeat.Blipping (System.Double beat, System.Single length) [0x000b8] in <6d2f5506bdb644ce89fc7390a1ea8702>:0 
  at HeavenStudio.Games.MrUpbeat.StartStepping (System.Double beat, System.Single length) [0x00024] in <6d2f5506bdb644ce89fc7390a1ea8702>:0 
  at HeavenStudio.Games.Loaders.AgbUpbeatLoader+<>c__DisplayClass0_0.<AddGame>b__0 () [0x00018] in <6d2f5506bdb644ce89fc7390a1ea8702>:0 
  at HeavenStudio.EventCaller.CallPreEvent (Jukebox.RiqEntity entity) [0x00050] in <6d2f5506bdb644ce89fc7390a1ea8702>:0 
== Playthrough statistics of New Remix (played at 2023-08-08 00:06:13) ==
Average input offset for playthrough: 0ms
Accuracy for playthrough:  20.00
Cleared 0 sections out of 0
Skill Star not collected
Playing at 0
== Playthrough statistics of New Remix (played at 2023-08-08 00:06:35) ==
Average input offset for playthrough: 0ms
Accuracy for playthrough:  100.00
Cleared 0 sections out of 0
Skill Star not collected
Perfect Clear!
Playing at 0
== Playthrough statistics of New Remix (played at 2023-08-08 00:06:43) ==
Average input offset for playthrough: 0ms
Accuracy for playthrough:  100.00
Cleared 0 sections out of 0
Skill Star not collected
Perfect Clear!
Disconnecting JoyShocks...
Clearing RIQ Cache...
Setting up 2 worker threads for Enlighten.
Memory Statistics:
[ALLOC_TEMP_TLS] TLS Allocator
  StackAllocators : 
    [ALLOC_TEMP_MAIN]
      Peak usage frame count: [0-1.0 KB]: 36 frames, [8.0 KB-16.0 KB]: 1 frames, [16.0 KB-32.0 KB]: 2048 frames, [32.0 KB-64.0 KB]: 1 frames, [64.0 KB-128.0 KB]: 1 frames, [128.0 KB-256.0 KB]: 1 frames, [256.0 KB-0.5 MB]: 1 frames, [0.5 MB-1.0 MB]: 1 frames, [2.0 MB-4.0 MB]: 1 frames
      Initial Block Size 4.0 MB
      Current Block Size 4.0 MB
      Peak Allocated Bytes 2.9 MB
      Overflow Count 0
    [ALLOC_TEMP_Loading.AsyncRead]
      Initial Block Size 64.0 KB
      Current Block Size 64.0 KB
      Peak Allocated Bytes 60.8 KB
      Overflow Count 0
    [ALLOC_TEMP_Loading.PreloadManager]
      Initial Block Size 256.0 KB
      Current Block Size 504.0 KB
      Peak Allocated Bytes 453.9 KB
      Overflow Count 19
    [ALLOC_TEMP_Background Job.Worker 8]
      Initial Block Size 32.0 KB
      Current Block Size 32.0 KB
      Peak Allocated Bytes 54 B
      Overflow Count 0
    [ALLOC_TEMP_Background Job.Worker 9]
      Initial Block Size 32.0 KB
      Current Block Size 32.0 KB
      Peak Allocated Bytes 54 B
      Overflow Count 0
    [ALLOC_TEMP_Job.Worker 0]
      Initial Block Size 256.0 KB
      Current Block Size 256.0 KB
      Peak Allocated Bytes 29.5 KB
      Overflow Count 0
    [ALLOC_TEMP_Background Job.Worker 10]
      Initial Block Size 32.0 KB
      Current Block Size 32.0 KB
      Peak Allocated Bytes 54 B
      Overflow Count 0
    [ALLOC_TEMP_Background Job.Worker 14]
      Initial Block Size 32.0 KB
      Current Block Size 32.0 KB
      Peak Allocated Bytes 54 B
      Overflow Count 0
    [ALLOC_TEMP_Background Job.Worker 6]
      Initial Block Size 32.0 KB
      Current Block Size 32.0 KB
      Peak Allocated Bytes 54 B
      Overflow Count 0
    [ALLOC_TEMP_ReadCacheIndex]
      Initial Block Size 64.0 KB
      Current Block Size 64.0 KB
      Peak Allocated Bytes 0.5 KB
      Overflow Count 0
    [ALLOC_TEMP_Background Job.Worker 12]
      Initial Block Size 32.0 KB
      Current Block Size 32.0 KB
      Peak Allocated Bytes 54 B
      Overflow Count 0
    [ALLOC_TEMP_EnlightenWorker] x 2
      Initial Block Size 64.0 KB
      Current Block Size 64.0 KB
      Peak Allocated Bytes 54 B
      Overflow Count 0
    [ALLOC_TEMP_Background Job.Worker 15]
      Initial Block Size 32.0 KB
      Current Block Size 32.0 KB
      Peak Allocated Bytes 54 B
      Overflow Count 0
    [ALLOC_TEMP_Background Job.Worker 1]
      Initial Block Size 32.0 KB
      Current Block Size 32.0 KB
      Peak Allocated Bytes 54 B
      Overflow Count 0
    [ALLOC_TEMP_Background Job.Worker 2]
      Initial Block Size 32.0 KB
      Current Block Size 32.0 KB
      Peak Allocated Bytes 54 B
      Overflow Count 0
    [ALLOC_TEMP_Background Job.Worker 7]
      Initial Block Size 32.0 KB
      Current Block Size 32.0 KB
      Peak Allocated Bytes 54 B
      Overflow Count 0
    [ALLOC_TEMP_UnityGfxDeviceWorker]
      Initial Block Size 256.0 KB
      Current Block Size 256.0 KB
      Peak Allocated Bytes 6.8 KB
      Overflow Count 0
    [ALLOC_TEMP_AssetGarbageCollectorHelper] x 3
      Initial Block Size 64.0 KB
      Current Block Size 64.0 KB
      Peak Allocated Bytes 0 B
      Overflow Count 0
    [ALLOC_TEMP_Background Job.Worker 5]
      Initial Block Size 32.0 KB
      Current Block Size 32.0 KB
      Peak Allocated Bytes 54 B
      Overflow Count 0
    [ALLOC_TEMP_Background Job.Worker 13]
      Initial Block Size 32.0 KB
      Current Block Size 32.0 KB
      Peak Allocated Bytes 54 B
      Overflow Count 0
    [ALLOC_TEMP_Job.Worker 1]
      Initial Block Size 256.0 KB
      Current Block Size 256.0 KB
      Peak Allocated Bytes 29.5 KB
      Overflow Count 0
    [ALLOC_TEMP_Job.Worker 2]
      Initial Block Size 256.0 KB
      Current Block Size 256.0 KB
      Peak Allocated Bytes 29.5 KB
      Overflow Count 0
    [ALLOC_TEMP_Background Job.Worker 3]
      Initial Block Size 32.0 KB
      Current Block Size 32.0 KB
      Peak Allocated Bytes 54 B
      Overflow Count 0
    [ALLOC_TEMP_Background Job.Worker 11]
      Initial Block Size 32.0 KB
      Current Block Size 32.0 KB
      Peak Allocated Bytes 54 B
      Overflow Count 0
    [ALLOC_TEMP_Background Job.Worker 0]
      Initial Block Size 32.0 KB
      Current Block Size 32.0 KB
      Peak Allocated Bytes 1.7 KB
      Overflow Count 1
    [ALLOC_TEMP_Background Job.Worker 4]
      Initial Block Size 32.0 KB
      Current Block Size 32.0 KB
      Peak Allocated Bytes 54 B
      Overflow Count 0
    [ALLOC_TEMP_BatchDeleteObjects]
      Initial Block Size 64.0 KB
      Current Block Size 64.0 KB
      Peak Allocated Bytes 54 B
      Overflow Count 0
[ALLOC_DEFAULT] Dual Thread Allocator
  Peak main deferred allocation count 826
    [ALLOC_BUCKET]
      Large Block size 4.0 MB
      Used Block count 1
      Peak Allocated bytes 3.5 MB
    [ALLOC_DEFAULT_MAIN]
      Peak usage frame count: [8.0 MB-16.0 MB]: 39 frames, [16.0 MB-32.0 MB]: 2052 frames
      Requested Block Size 16.0 MB
      Peak Block count 2
      Peak Allocated memory 18.2 MB
      Peak Large allocation bytes 0 B
    [ALLOC_DEFAULT_THREAD]
      Peak usage frame count: [16.0 MB-32.0 MB]: 2091 frames
      Requested Block Size 16.0 MB
      Peak Block count 1
      Peak Allocated memory 19.9 MB
      Peak Large allocation bytes 16.0 MB
[ALLOC_TEMP_JOB_1_FRAME]
  Initial Block Size 2.0 MB
  Used Block Count 1
  Overflow Count (too large) 0
  Overflow Count (full) 0
[ALLOC_TEMP_JOB_2_FRAMES]
  Initial Block Size 2.0 MB
  Used Block Count 1
  Overflow Count (too large) 0
  Overflow Count (full) 0
[ALLOC_TEMP_JOB_4_FRAMES (JobTemp)]
  Initial Block Size 2.0 MB
  Used Block Count 2
  Overflow Count (too large) 0
  Overflow Count (full) 0
[ALLOC_TEMP_JOB_ASYNC (Background)]
  Initial Block Size 1.0 MB
  Used Block Count 5
  Overflow Count (too large) 0
  Overflow Count (full) 0
[ALLOC_GFX] Dual Thread Allocator
  Peak main deferred allocation count 2
    [ALLOC_BUCKET]
      Large Block size 4.0 MB
      Used Block count 1
      Peak Allocated bytes 3.5 MB
    [ALLOC_GFX_MAIN]
      Peak usage frame count: [1.0 MB-2.0 MB]: 38 frames, [2.0 MB-4.0 MB]: 1 frames, [4.0 MB-8.0 MB]: 2052 frames
      Requested Block Size 16.0 MB
      Peak Block count 1
      Peak Allocated memory 5.5 MB
      Peak Large allocation bytes 0 B
    [ALLOC_GFX_THREAD]
      Peak usage frame count: [128.0 KB-256.0 KB]: 39 frames, [256.0 KB-0.5 MB]: 1 frames, [0.5 MB-1.0 MB]: 2051 frames
      Requested Block Size 16.0 MB
      Peak Block count 1
      Peak Allocated memory 0.5 MB
      Peak Large allocation bytes 0 B
[ALLOC_CACHEOBJECTS] Dual Thread Allocator
  Peak main deferred allocation count 749
    [ALLOC_BUCKET]
      Large Block size 4.0 MB
      Used Block count 1
      Peak Allocated bytes 3.5 MB
    [ALLOC_CACHEOBJECTS_MAIN]
      Peak usage frame count: [8.0 MB-16.0 MB]: 629 frames, [16.0 MB-32.0 MB]: 1462 frames
      Requested Block Size 4.0 MB
      Peak Block count 2
      Peak Allocated memory 24.0 MB
      Peak Large allocation bytes 17.5 MB
    [ALLOC_CACHEOBJECTS_THREAD]
      Peak usage frame count: [1.0 MB-2.0 MB]: 545 frames, [2.0 MB-4.0 MB]: 41 frames, [4.0 MB-8.0 MB]: 23 frames, [8.0 MB-16.0 MB]: 1482 frames
      Requested Block Size 4.0 MB
      Peak Block count 4
      Peak Allocated memory 11.6 MB
      Peak Large allocation bytes 0 B
[ALLOC_TYPETREE] Dual Thread Allocator
  Peak main deferred allocation count 0
    [ALLOC_BUCKET]
      Large Block size 4.0 MB
      Used Block count 1
      Peak Allocated bytes 3.5 MB
    [ALLOC_TYPETREE_MAIN]
      Peak usage frame count: [16.0 KB-32.0 KB]: 39 frames, [32.0 KB-64.0 KB]: 590 frames, [256.0 KB-0.5 MB]: 1462 frames
      Requested Block Size 2.0 MB
      Peak Block count 1
      Peak Allocated memory 329.6 KB
      Peak Large allocation bytes 0 B
    [ALLOC_TYPETREE_THREAD]
      Peak usage frame count: [4.0 KB-8.0 KB]: 37 frames, [8.0 KB-16.0 KB]: 588 frames, [128.0 KB-256.0 KB]: 1466 frames
      Requested Block Size 2.0 MB
      Peak Block count 1
      Peak Allocated memory 167.1 KB
      Peak Large allocation bytes 0 B
```

### RIQ file

[remix_bug_forgetful_upbeat.riq.zip](https://github.com/RHeavenStudio/HeavenStudio/files/12288290/remix_bug_forgetful_upbeat.riq.zip)

This RIQ if you play it a few times should stop blipping at all, not start to step on one of the two tries, but if you wait 10 or so seconds it might start briefly.
(This is what I observed)

### Additional context

For me, sometimes it works once then breaks when I try again, or when I move the blocks it uses the old position once and then stops working completely.

I understand that the game is not feature-complete (looking at comments in the code saying so), although this should still be noted for whenever it might be finished. I hope it's something you can reproduce just because that would mean I'm not completely crazy.

**Comment by AstrlJelly** [comment]
im redoing my mr. upbeat rework right now, and i'll be testing it much more rigorously, so that it should be foolproof. the way i coded everything before was bad and made everything way too inconsistent.

**Comment by blank3times** [comment]
Fixed by #525 and/or #531
